File: packages/server/src/routes/retros.ts
import { Router } from "express";
import { eq, asc } from "drizzle-orm";
import { db } from "../db";
import { retros, retroColumns } from "../db/schema";
import { generateId } from "../lib/id";
import type { CreateRetroPayload } from "../types";

const router = Router();

// POST /api/retros — crear retro
router.post("/", (req, res) => {
  const { title, timerSeconds, votesPerPerson, facilitatorEmail, roomId, columns } =
    req.body as CreateRetroPayload;

  if (!title?.trim() || !facilitatorEmail || !columns?.length) {
    res.status(400).json({ error: "Faltan campos requeridos" });
    return;
  }

  const id = generateId();
  const now = Date.now();

  db.insert(retros)
    .values({
      id,
      title: title.trim(),
      facilitatorEmail,
      roomId: roomId ?? null,
      timerSeconds: timerSeconds ?? 300,
      votesPerPerson: votesPerPerson ?? 5,
      phase: "waiting",
      createdAt: now,
    })
    .run();

  columns.forEach((col, idx) => {
    db.insert(retroColumns)
      .values({
        id: generateId(),
        retroId: id,
        title: col.title,
        emoji: col.emoji ?? null,
        position: idx + 1,
      })
      .run();
  });

  res.status(201).json({ id });
});

// GET /api/retros/:id — info de la retro (para unirse)
router.get("/:id", (req, res) => {
  const retro = db.select().from(retros).where(eq(retros.id, req.params.id)).get();
  if (!retro) {
    res.status(404).json({ error: "Retrospectiva no encontrada" });
    return;
  }

  const columns = db
    .select()
    .from(retroColumns)
    .where(eq(retroColumns.retroId, retro.id))
    .orderBy(asc(retroColumns.position))
    .all();

  res.json({ retro, columns });
});

export default router;

File: packages/server-python/routes/retros.py
from typing import Optional

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel

from db import fetchone, fetchall, execute, generate_id, row_to_camel, now_ms

router = APIRouter()


class ColumnSpec(BaseModel):
    title: str
    emoji: Optional[str] = None


class CreateRetroBody(BaseModel):
    title: str
    timerSeconds: int = 300
    votesPerPerson: int = 5
    facilitatorEmail: str
    roomId: Optional[str] = None
    columns: list[ColumnSpec]


@router.post("", status_code=201)
async def create_retro(body: CreateRetroBody):
    if not body.title.strip() or not body.facilitatorEmail or not body.columns:
        raise HTTPException(status_code=400, detail="Faltan campos requeridos")

    retro_id = generate_id()
    ts = now_ms()

    execute(
        "INSERT INTO retros (id, title, facilitator_email, room_id, timer_seconds, "
        "votes_per_person, phase, created_at) VALUES (?, ?, ?, ?, ?, ?, 'waiting', ?)",
        (
            retro_id,
            body.title.strip(),
            body.facilitatorEmail,
            body.roomId,
            body.timerSeconds,
            body.votesPerPerson,
            ts,
        ),
    )

    for idx, col in enumerate(body.columns):
        execute(
            "INSERT INTO retro_columns (id, retro_id, title, emoji, position) VALUES (?, ?, ?, ?, ?)",
            (generate_id(), retro_id, col.title, col.emoji, idx + 1),
        )

    return {"id": retro_id}


@router.get("/{retro_id}")
async def get_retro(retro_id: str):
    retro = fetchone("SELECT * FROM retros WHERE id = ?", (retro_id,))
    if not retro:
        raise HTTPException(status_code=404, detail="Retrospectiva no encontrada")

    columns = fetchall(
        "SELECT * FROM retro_columns WHERE retro_id = ? ORDER BY position ASC", (retro_id,)
    )

    return {
        "retro": row_to_camel(retro),
        "columns": [row_to_camel(c) for c in columns],
    }
